### ISSUE TITLE
Add capacity to have particles die and split

### ISSUE THREAD

**Issue by erikvansebille** [OceanPARCELS/parcels#14]
It would be great if parcels could randomly die and split over their trajectories, simulating the evolution of microbes in the ocean. This was done within CMS for the paper at http://science.sciencemag.org/content/345/6202/1346.abstract

Not sure if we want this capacity already for the GMD manuscript

**Comment by mlange05** [comment]
This really poses the question of data structures, especially for the C/JIT mode. To implement this, I think we need the following:
* Add particle IDs and change I/O format to an [indexed ragged array representation](http://cfconventions.org/cf-conventions/v1.6.0/cf-conventions.html#_indexed_ragged_array_representation).
* Add `ParticleSet.add/remove()` API and possibly some nice `Particle.merge(weights)` routines for the Python mode. Since the underlying representation here is a Python list we only need to look at what the API should look like to make it simple to use in custom kernels and from outside the loop.
* For C/JIT we need to change the underlying data structure. We currently use pre-allocated numpy arrays, but we probably want some type of smart linked-list. This sounds like something really important that we need to do anyway at some point, but might have a significant impact on performance and other features, like parallelism (see issue #17).

So yes, I think, we want this for the GMD paper.

**Comment by mlange05** [comment]
Ok, after trying a whole bunch of different things I have come to the conclusion that changing to underlying data structures to C++ `std::vector` and/or `std::list` types is a) harder than expected and b) not necessarily what we want right now. To explain, I spend a lot of time trying to do the following things:
* Use the [codepy.bpl](https://documen.tician.de/codepy/jit.html#module-codepy.bpl) to utilise boost.python, one of the standard Python-C++ interoperability packages, to do our JIT compilation. I got as far as being able to build and mange C++ `vector<Particle>` objects, only to discover that boost.python does not expose the underlying C objects. This means that I found no way to perform our C/C++ execution loop without doing function pointer juggling in the inner loops. 
* Next, I tried wrapping `std::vector` myself and even got this to work (see branch `cpp-dynamic-lists`) completely. However, it turns out that the Python-C++ bindings I used for setting individual particle fields/attributes are significantly slower than the ones we are currently using from NumPy, so I don't think this is an ideal solution either.

A further point on the data structure argument is that it really boils down to array/vector types with fast execution due to alignment vs. lists/linked-lists with fast insertion/deletion, but potentially slow kernel execution. Ultimately, I believe, a combination of the two, ie. lists of fixed-size buffer arrays with some custom insertion/deletion management, will provide the best of both worlds. However, this sounds like an optimisation for a later stage in the project and it is not clear to me that we necessarily require C++ types of the form `std::list<std::vector<Particle>>`, or whether we can implement this type of scheme using Python lists and numpy arrays.

So, for now I think I will abandon any further attempts at using C++ data structures and focus on extending the `ParticleSet` interface with insertion and deletion methods, even though they will probably be fairly slow for large particle arrays. We can revisit optimisation ideas once we have the core functionality and test cases that expose the shortcoming more clearly.
